Index: smn_analysis.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>#!/usr/env python\n\n\"\"\" Read CRAM files\"\"\"\n\"\"\" The cram files contains the data of SMN mutations diagnosed samples. \"\"\"\nimport pysam\nHUMAN_G = '/Users/rakshya/SMA.2022/Homo_sapiens_assembly38.fasta'\nCRAM_F = '/Users/rakshya/SMA.2022/MAAC057.cram'\ncram = pysam.AlignmentFile(CRAM_F, 'rc', reference_filename=HUMAN_G)\nSMN1 = cram.count(region=\"chr5:70924941-70953012\")\nprint(\"SMN1 : \", SMN1)\n#Output - SMN1 :  7088\nSMN2 = cram.count(region=\"chr5:70049523-70077595\")\nprint(\"SMN2 : \", SMN2)\n#Output - SMN2 :  7438\ncram.close()\n\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/smn_analysis.py b/smn_analysis.py
--- a/smn_analysis.py	(revision d6d5e0612497065711187b9f50080d20ee68bb06)
+++ b/smn_analysis.py	(date 1656083343443)
@@ -1,4 +1,4 @@
-#!/usr/env python
+#!/usr/env python3
 
 """ Read CRAM files"""
 """ The cram files contains the data of SMN mutations diagnosed samples. """
